### ISSUE TITLE
При повторной загрузке плейлиста, уже имеющиеся файлы дублируются

### ISSUE THREAD

**Issue by novikovsergey** [egoroof/yandex-music-fisher#28]
Привет, есть ли возможность на уровне екстеншена перед созданием и загрузкой файла проверять его наличия, допустим по названию и сделать настройку, которая бы отключала скачивания таких файлов?

Возможный кейс: Дополнил плейлист, через некоторое время хочу докачать недостающие композиции и приходится весь плейлист перекачивать.

 Спасибо за ответ!

**Comment by faergeek** [comment]
+1

**Comment by egoroof** [comment]
Такая задача уже есть у меня в todo :)
По поводу реализации - расширение доступ к файлам на диске не имеет, поэтому оно будет запоминать скачанные плейлисты и треки, а при повторном скачивании исключать их.

**Comment by faergeek** [comment]
А что если предоставить пользователю возможность выбирать из списка песни, которые он хочет скачать?

**Comment by egoroof** [comment]
Можно ага и у скачанных ранее треков галочка будет снята.

**Comment by novikovsergey** [comment]
Да, хорошие варианты, Ток ещё наследоваться надо не забыть от источника, чтобы была возможность альбом целиком скачать, даже если этот трек из альбома ранее скачан был из плейлиста.

**Comment by faergeek** [comment]
@egoroof уже есть какие-то наработки? Я был бы рад помочь с реализацией.

**Comment by egoroof** [comment]
@faergeek не, пока совсем не до этого - если хочешь сделать pull request, буду рад!

**Comment by Gravity89** [comment]
Я не уверен, но может еще хэш хранить? Или длительность трека. Чтобы каверы не блочить. Или может их API дает доступ к ID трека в базе?

**Comment by egoroof** [comment]
@Gravity89 у каждого трека свой уникальный идентфикатор. Он есть в URL страниц, например music.yandex.ru/album/2986409/track/**2160949** здесь 2160949 - ID.

**Comment by Gravity89** [comment]
Это ID трека в альбоме. Если я скажем найду 1:1 такой же трек в другом альбоме, то там будет другой ID же?

**Comment by egoroof** [comment]
Неа, это уникальный ID. Причём в другом альбоме может быть трек с точно таким же ID.
Вот пример работы функции, которая по ID трека создаёт ссылки на все альбомы с этим треком:
![2015-12-28 15 55 32](https://cloud.githubusercontent.com/assets/2605831/12015718/85e28688-ad7b-11e5-9f94-617363e010e2.png)

**Comment by MihailovAlexander** [comment]
Очень жду эту функцию. Есть плейлист, который постоянно обновляется, вот только новый добавленные бы докачивать, Спасибо!

**Comment by Barefooter** [comment]
Да, полезная функция была бы. Сейчас каждый раз заново все качаю.

9 июня 2016 г., 9:42 пользователь MihailovAlexander <
notifications@github.com> написал:

> Очень жду эту функцию. Есть плейлист, который постоянно обновляется, вот
> только новый добавленные бы докачивать, Спасибо!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/egoroof/yandex-music-fisher/issues/28#issuecomment-224812785>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAoJpdnWqxyS-lFKYkwwkgfQ1GQfFhekks5qJ7XAgaJpZM4F6cOP>
> .
>



-- 
Станислав Агарков

**Comment by Gravity89** [comment]
Отголосок #79 
Забавно. Может кто подымет внешнюю БД для хранения уже скачанных треков в разрезе пользователей? XD
Тогда можно плагином посылать запрос к внешней API "User:TrackID" и не качать что уже в базе. Только надо еще очистку или принудительное скачивание придумать... 
Ну или писать все в txt на Яндекс Диске (или любом другом облаке) под своим логином и паролем.

**Comment by gubenkoved** [comment]
+1 за такую функциональность -- очень полезно в случае больших (300+), обновляемых плейлистов.

Не уверен насколько релевантно, но кажется Chrome поддерживает API для доступа к локальной ФС -- https://developer.chrome.com/apps/fileSystem.

**Comment by faergeek** [comment]
@gubenkoved это API для приложений, а не расширений. А тут у нас расширение

**Comment by gubenkoved** [comment]
@faergeek ага, вижу теперь, спасибо!

**Comment by gubenkoved** [comment]
@faergeek а что насчет вот такого подхода -- chrome.downloads API поддерживает поиск по загрузкам и кажется может подсказать существует ли файл на диске. Понятно, что как только пользователь очищает историю загрузок такой способ уже скорее всего не сработает, но всё же это может покрыть большинство кейсов...

> Whether the downloaded file still exists. This information may be out of date because Chrome does not automatically watch for file removal. Call search() in order to trigger the check for file existence. When the existence check completes, if the file has been deleted, then an onChanged event will fire. Note that search() does not wait for the existence check to finish before returning, so results from search() may not accurately reflect the file system. Also, search() may be called as often as necessary, but will not check for file existence any more frequently than once every 10 seconds.

https://developer.chrome.com/extensions/downloads#method-search
https://developer.chrome.com/extensions/downloads#type-DownloadItem

**Comment by faergeek** [comment]
@gubenkoved для меня не сработает, я загрузки всегда чищу)) Привычка. Я бы и реализовывать такое не стал, хотя бы только поэтому.

**Comment by gubenkoved** [comment]
Ага, вижу что очищаются загрузки -- можно спросить чем это обосновано?

**Comment by faergeek** [comment]
@gubenkoved я сейчас не про расширение говорил, а про то как я браузером пользуюсь :-). Тут я, честно говоря, до кода так и не добрался. Просто нет идеи как это реализовать. Если только отдельное приложение, но это вообще не решение. Неудобно для пользователя.

**Comment by gubenkoved** [comment]
@faergeek да, не всем подойдет конечно, но потенциально можно сделать это поведение конфигурируемым, вроде как продвинутая настройка "_Try_ not to overwrite existing files", при включении которой пользователю сразу говорится об ограничениях такой функциональности.

Спасибо за ответы!
